File: SkyEditor.Core.Windows/Windows/EnvironmentPaths.vb
Imports System.Deployment.Application
Imports System.IO
Imports System.Reflection
Imports SkyEditor.Core.Extensions

Namespace Windows
    Public Class EnvironmentPaths
        ''' <summary>
        ''' Gets the directory used to store extensions.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetExtensionDirectory() As String
            Return Path.Combine(GetRootResourceDirectory, "Extensions")
        End Function

        Public Shared Function GetPluginsExtensionDirectory As String
            Return Path.Combine(GetExtensionDirectory, (New PluginExtensionType).InternalName)
        End Function

        ''' <summary>
        ''' Combines the given path with your plugin's resource directory.
        ''' </summary>
        ''' <param name="Path"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Shared Function GetResourceName(Path As String) As String
            Return GetResourceName(Path, Assembly.GetCallingAssembly.GetName.Name)
        End Function

        Public Shared Function GetResourceName(resoucePath As String, pluginName As String, Optional ThrowIfCantCreateDirectory As Boolean = False) As String
            Dim fullPath = Path.Combine(GetResourceDirectory(pluginName), resoucePath)
            Dim baseDir = Path.GetDirectoryName(fullPath)
            Try
                If Not Directory.Exists(baseDir) Then
                    Directory.CreateDirectory(baseDir)
                End If
            Catch ex As UnauthorizedAccessException
                If ThrowIfCantCreateDirectory Then
                    Throw ex
                End If
            End Try
            Return fullPath
        End Function
        ''' <summary>
        ''' Returns your plugin's resource directory as managed by Sky Editor.
        ''' It will be created if it does not exist.
        ''' </summary>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Shared Function GetResourceDirectory() As String
            Return GetResourceDirectory(Assembly.GetCallingAssembly.GetName.Name)
        End Function

        ''' <summary>
        ''' Gets the resource directory for the plugin with the given assembly name.
        ''' </summary>
        ''' <param name="AssemblyName"></param>
        ''' <param name="ThrowIfCantCreateDirectory"></param>
        ''' <returns></returns>
        Public Shared Function GetResourceDirectory(AssemblyName As String, Optional ThrowIfCantCreateDirectory As Boolean = False) As String
            Dim baseDir = Path.Combine(GetRootResourceDirectory, "Extensions", "Plugins", "Development", AssemblyName)
            If Directory.Exists(baseDir) Then
                Return baseDir
            ElseIf Directory.Exists(Path.Combine(Environment.CurrentDirectory, AssemblyName)) Then
                Return Path.Combine(Environment.CurrentDirectory, AssemblyName)
            Else
                Try
                    Directory.CreateDirectory(baseDir)
                Catch ex As UnauthorizedAccessException
                    If ThrowIfCantCreateDirectory Then
                        Throw ex
                    End If
                End Try
                Return baseDir
            End If
        End Function

        ''' <summary>
        ''' Returns a the path of the root resource directory and creates it if it doesn't exist.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetRootResourceDirectory(Optional ThrowIfCantCreateDirectory As Boolean = False) As String
            Dim d As String
            If ApplicationDeployment.IsNetworkDeployed Then
                'I'm choosing not to verify if the folder exists because I'm already going to check below.
                d = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.Create)
            Else
                d = Path.Combine(Environment.CurrentDirectory, "Resources")
                If Not Directory.Exists(d) Then
                    Try
                        Directory.CreateDirectory(d)
                    Catch ex As UnauthorizedAccessException
                        If ThrowIfCantCreateDirectory Then
                            Throw ex
                        End If
                    End Try
                End If
            End If
            Return d
        End Function

        ''' <summary>
        ''' Gets the full path for the application's settings file.
        ''' </summary>
        ''' <returns></returns>
        Public Shared Function GetSettingsFilename() As String
            Return Path.Combine(EnvironmentPaths.GetRootResourceDirectory, "settings.json")
        End Function

    End Class
End Namespace

File: SkyEditor.Core/Utilities/EnvironmentPaths.cs
using SkyEditor.Core.Extensions;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace SkyEditor.Core.Utilities
{
    /// <summary>
    /// Determines various file paths
    /// </summary>
    public static class EnvironmentPaths
    {
        /// <summary>
        /// Gets the root resource directory, where plugins, settings, etc. are stored.
        /// </summary>
        /// <param name="throwIfCreateFails">Whether or not to throw an exception if the directory cannot be created</param>
        /// <returns>The path of the root resource directory</returns>
        public static string GetRootResourceDirectory(bool throwIfCreateFails = false)
        {
#if NET462
            // Store things in local app data
            if (System.Deployment.Application.ApplicationDeployment.IsNetworkDeployed)
            {
                return Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.Create);
            }
#endif
            // Store things in working directory
            var resourcesDir = Path.Combine(Directory.GetCurrentDirectory(), "Resources");
            if (!Directory.Exists(resourcesDir))
            {
                try
                {
                    Directory.CreateDirectory(resourcesDir);
                }
                catch (UnauthorizedAccessException)
                {
                    if (throwIfCreateFails)
                    {
                        throw;
                    }                    
                }
            }
            return resourcesDir;
        }

        /// <summary>
        /// Gets the path of the settings file
        /// </summary>
        public static string GetSettingsFilename()
        {
            return Path.Combine(GetRootResourceDirectory(), "settings.json");
        }

        /// <summary>
        /// Gets directory in which extensions are stored
        /// </summary>
        public static string GetExtensionsDirectory()
        {
            return Path.Combine(GetRootResourceDirectory(), "Extensions");
        }

        /// <summary>
        /// Gets the Plugins extension directory
        /// </summary>
        public static string GetPluginsExtensionDirectory()
        {
            return Path.Combine(GetExtensionsDirectory(), PluginExtensionType.PluginsInternalName);
        }

#if NET462

        /// <summary>
        /// Gets the directory for a plugin with the given assembly name
        /// </summary>
        [Obsolete] public static string GetResourceDirectory(string assemblyName, bool throwIfCreateFails = false)
        {
            var devDir = Path.Combine(GetRootResourceDirectory(), "Extensions", "Plugins", "Development", assemblyName);
            var pluginDirectory = Path.Combine(Directory.GetCurrentDirectory(), assemblyName);
            if (Directory.Exists(devDir))
            {
                return devDir;
            }
            else if (Directory.Exists(pluginDirectory))
            {
                return pluginDirectory;
            }
            else
            {
                try
                {
                    Directory.CreateDirectory(devDir);
                }
                catch (Exception)
                {
                    if (throwIfCreateFails)
                    {
                        throw;
                    }
                }
                return devDir;
            }
        }

        /// <summary>
        /// Returns the calling plugin's resource directory
        /// </summary>
        /// <returns></returns>
        [Obsolete] public static string GetResourceDirectory()
        {
            return GetResourceDirectory(Assembly.GetCallingAssembly().GetName().Name);
        }

        [Obsolete] public static string GetResourceName(string resourcePath, string pluginName, bool throwIfCreateFails = false)
        {
            var fullPath = Path.Combine(GetResourceDirectory(pluginName), resourcePath);
            var baseDir = Path.GetDirectoryName(fullPath);
            try
            {
                if (!Directory.Exists(baseDir))
                {
                    Directory.CreateDirectory(baseDir);
                }
            }
            catch (Exception)
            {
                if (throwIfCreateFails)
                {
                    throw;
                }
            }
            return fullPath;
        }

        /// <summary>
        /// Combine's the given path with the calling plugin's resource directory
        /// </summary>
        [Obsolete] public static string GetResourceName(string path)
        {
            return GetResourceName(path, Assembly.GetCallingAssembly().GetName().Name);
        }
#endif
    }
}
